pyro/ops/jit.py

```diff
@@ -0,0 +1,80 @@
+import weakref
+import torch
+
+import pyro
+import pyro.poutine as poutine
+
+
+class compile(object):
+    """
+    Drop-in replacement for :func:`torch.jit.compile` that works with
+    Pyro functions that call :func:`pyro.param`.
+
+    The actual compilation artifact is stored in ``self.compiled``.
+    Call diagnostic methods on this attribute.
+
+    Example::
+
+        def model(x):
+            scale = pyro.param("scale", torch.tensor(0.5), constraint=constraints.positive)
+            return pyro.sample("y", dist.Normal(x, scale))
+
+        @pyro.ops.jit.compile(nderivs=1)
+        def model_log_prob_fn(x, y):
+            cond_model = pyro.condition(model, data={"y": y})
+            tr = pyro.poutine.trace(cond_model).get_trace(x)
+            return tr.log_prob_sum()
+    """
+    def __init__(self, **kwargs):
+        self._kwargs = kwargs
+        self.compiled = None
+        self._param_names = None
+
+    def __call__(self, fn):
+
+        assert self.compiled is None, \
+            "Cannot use a compile instance on more than one function"
+
+        def wrapper(*args, **kwargs):
+            # if first time
+            if self.compiled is None:
+                # param capture
+                with poutine.block():
+                    with poutine.trace(param_only=True) as first_param_capture:
+                        fn(*args, **kwargs)
+
+                self._param_names = list(set(first_param_capture.trace.nodes.keys()))
+
+                weakself = weakref.ref(self)
+
+                @torch.jit.compile(**self._kwargs)
+                def compiled(unconstrained_params, *args):
+                    self = weakself()
+                    constrained_params = {name: pyro.param(name, unconstrained_param)
+                                          for name, unconstrained_param
+                                          in zip(self._param_names,
+                                                 unconstrained_params)}
+
+                    return poutine.replay(
+                        fn, params=constrained_params)(*args, **kwargs)
+
+                self.compiled = compiled
+
+            param_list = [pyro.param(name).unconstrained()
+                          for name in self._param_names]
+
+            with poutine.block(hide=self._param_names):
```

Comment:
Can you help me undersand why you're blocking and tracing here?

Reply:
I guess I wanted to avoid having `pyro.param` be called twice for params in `self._param_names`, but I'm not sure that matters?

Reply:
Ok, I see that it makes sense to capture new params, and if you're capturing new params, it makes sense to block. Thanks for explaining.